xapian-core/tests/unittest.cc

```diff
@@ -536,10 +536,22 @@ static bool test_tostring1()
     TEST_EQUAL(str(-1), "-1");
     TEST_EQUAL(str(-9), "-9");
     TEST_EQUAL(str(-10), "-10");
+    TEST_EQUAL(str(0x7f), "127");
+    TEST_EQUAL(str(-0x80), "-128");
+    TEST_EQUAL(str(0x7fff), "32767");
+    // test for lvalue with minimum value for short type.
+    short temp = -32768;
+    TEST_EQUAL(str(temp), "-32768");
+
     TEST_EQUAL(str(0xffffffff), "4294967295");
     TEST_EQUAL(str(0x7fffffff), "2147483647");
     TEST_EQUAL(str(0x7fffffffu), "2147483647");
-    TEST_EQUAL(str(-0x7fffffff), "-2147483647");
+
+    // test for lvalue with minimum value for int type.
+    int temp1 = -2147483648;
+    TEST_EQUAL(str(temp1), "-2147483648");
```

Comment:
I don't think it's the lvalue-ness that's the issue here - rather a slightly surprising behaviour of the types of integer constants - consider:

```
olly@gemse:~$ cat constant-type-probes.cc 
#include <iostream>
using namespace std;

#define DEF_F(T) void f(T v) { cout << #T "(" << v << ")\n"; }

DEF_F(short)
DEF_F(int)
DEF_F(long)
DEF_F(long long)

int main() {
    f(32767);
    f(-32767);
    f(-32768);
    f(2147483647);
    f(-2147483647);
    f(-2147483648);
    f(int(-2147483648));
}
olly@gemse:~$ g++ -Wall -W -O2 constant-type-probes.cc
olly@gemse:~$ ./a.out
int(32767)
int(-32767)
int(-32768)
int(2147483647)
int(-2147483647)
long(-2147483648)
int(-2147483648)
```

The slightly odd thing here is that `-2147483648` is of type `long` when the value actually fits in an `int`.  I think this is because it's actually parsed as `-(2147483648)` and `2147483648` doesn't fit in an `int`, so has type `long`.  Then that's negated, and the negation preserves the type.

But rather than using temporary variables, we can just write such testcases with an explicit conversion to the type we want to test, e.g.:

```
TEST_EQUAL(str(int(-2147483648)), "-2147483648");
```